Feature preview: Bootspecs as a NixOS module with tests (#4)
DeterminateSystems/nixpkgs: nixos/doc/manual/development/bootspec.chapter.md

# Experimental feature: Bootspec {#sec-experimental-bootspec}

> Bootspec is a experimental feature, introduced in [RFC-0125 proposal](https://github.com/NixOS/nixpkgs/pull/172237) in order to standardize bootloader support

Reviewer: ```suggestion
Bootspec is a experimental feature, introduced in the [RFC-0125 proposal](https://github.com/NixOS/nixpkgs/pull/172237) in order to standardize bootloader support
```

Also, I wonder if that  should link directly to RFC 125, but have an aside that links to that "reference" implementation? i.e. `introduced in [RFC-0125](..../rfcs/pull/125) (the reference implementation can be found at nixpkgs/pull/172237)` or something like that.

Author: Can add reference to implementation

and advanced boot workflows such as SecureBoot and potentially more.

> You can enable bootspec generation through [`boot.bootspec.enable = true`](options.html#opt-boot.bootspec.enable), which will prompt a warning until RFC-0125 is officially merged.

Reviewer: ```suggestion
You can enable the creation of bootspec documents through [`boot.bootspec.enable = true`](options.html#opt-boot.bootspec.enable), which will prompt a warning until RFC-0125 is officially merged.
```

Since Nix and NixOS overload "generation" a lot, I tend to shy away from using it.

Also, it would be nice to link directly to the RFC every time it is mentioned.

Author: Okay for link

## Schema {#sec-experimental-bootspec-schema}

> Bootspec schema is versioned and validated against [a CUE schema file](https://cuelang.org/) which should considered as the source of truth for your applications.

Reviewer: ```suggestion
The bootspec schema is versioned and validated against [a CUE schema file](https://cuelang.org/) which should considered as the source of truth for your applications.
```

> You will find the latest version here: <https://github.com/NixOS/nixpkgs/tree/master/nixos/modules/system/activation/bootspec.cue>.

Reviewer: Is there an easy way to have this be a relative link? Will the md-to-docbook stuff be smart enough to handle that? I don't like the hardcoded URL since, as it stands now, it will 404 :P

Author: seems fine

## Extensions mechanism {#sec-experimental-bootspec-extensions}

> Bootspec cannot account for all usecases and some of them should not end up in the global document.

Reviewer: ```suggestion
Bootspec cannot account for all usecases and some of them should not end up in the global document.
```

Honestly, I think that you could drop "and some of them........" and it would still be fine, but that's not a strongly-held opinion.

Author: Alright can do

> For this purpose, Bootspec offers [a generic extension facility](options.html#opt-boot.bootspec.extensions) in the `extensions` field which can be used to inject any data needed for your usecases.

Reviewer: ```suggestion
For this purpose, Bootspec offers a generic extension facility [`boot.bootspec.extensions`](options.html#opt-boot.bootspec.extensions) which can be used to inject any data needed for your usecases.
```

Even though there's an example below, I'd rather have the full option linked here as well.

An example for SecureBoot is to get the Nix store path to `/etc/os-release` in order to bake it into a unified kernel image:

```nix
{ config, lib, ... }: {
  boot.bootspec.extensions = {
    "org.secureboot.osRelease" = config.environment.etc."os-release".source;
  };
}
```

> To reduce incompatibility and namespace clash, you **should** use your own namespace for your own extensions.

Reviewer: ```suggestion
To reduce incompatibility and prevent names from clashing between applications, it is **highly recommended** to use a unique namespace for your extensions.
```
